machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,593 @@
+
+	// lastTransitionTime is the time of the last update to the current status object.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+
+	// reason is the reason for the condition's last transition.  Reasons are CamelCase
+	Reason string `json:"reason,omitempty"`
+
+	// message provides additional information about the current condition.
+	// This is only to be consumed by humans.
+	Message string `json:"message,omitempty"`
+}
+
+// ControllerConfigStatusConditionType valid conditions of a machineconfigpool
+type ControllerConfigStatusConditionType string
+
+const (
+	TemplateContollerRunning   ControllerConfigStatusConditionType = "TemplateContollerRunning"
+	TemplateContollerCompleted ControllerConfigStatusConditionType = "TemplateContollerCompleted"
+	TemplateContollerFailing   ControllerConfigStatusConditionType = "TemplateContollerFailing"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ControllerConfigList is a list of ControllerConfig resources
+type ControllerConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []ControllerConfig `json:"items"`
+}
+
+// CustomResourceDefinition for MachineConfig
+// apiVersion: apiextensions.k8s.io/v1beta1
+// kind: CustomResourceDefinition
+// metadata:
+//   # name must match the spec fields below, and be in the form: <plural>.<group>
+//   name: machineconfigs.machineconfiguration.openshift.io
+// spec:
+//   # group name to use for REST API: /apis/<group>/<version>
+//   group: machineconfiguration.openshift.io
+//   # list of versions supported by this CustomResourceDefinition
+//   versions:
+//     - name: v1
+//       # Each version can be enabled/disabled by Served flag.
+//       served: true
+//       # One and only one version must be marked as the storage version.
+//       storage: true
+//   # either Namespaced or Cluster
+//   scope: Cluster
+//   names:
+//     # plural name to be used in the URL: /apis/<group>/<version>/<plural>
+//     plural: machineconfigs
+//     # singular name to be used as an alias on the CLI and for display
+//     singular: machineconfig
+//     # kind is normally the CamelCased singular type. Your resource manifests use this.
+//     kind: MachineConfig
+//     # shortNames allow shorter string to match your resource on the CLI
+//     shortNames:
+//     - mc
+
+// +genclient
+// +genclient:noStatus
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen=false
+
+// MachineConfig defines the configuration for a machine
+type MachineConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec MachineConfigSpec `json:"spec"`
+}
+
+// MachineConfigSpec is the for MachineConfig
+type MachineConfigSpec struct {
+	// OSImageURL specifies the remote location that will be used to
+	// fetch the OS.
+	OSImageURL string `json:"osImageURL"`
+	// Config is a Ignition Config object.
+	Config ignv2_2types.Config `json:"config"`
```

Comment:
do we completely control these external types? Usually embedding of foreign types is dangerous.

Reply:
This part is actually removed from the PR as we're only moving type definitions which do not depend on Ignition